norddrop/src/ffi/uni.rs

```diff
@@ -0,0 +1,482 @@
+    fn level(&self) -> crate::LogLevel;
+}
+
+pub trait KeyStore: Send + Sync {
+    fn on_pubkey(&self, peer: String) -> Option<Vec<u8>>;
+    fn privkey(&self) -> Vec<u8>;
+}
+
+pub trait FdResolver: Send + Sync {
+    fn on_fd(&self, content_uri: String) -> Option<i32>;
+}
+
+pub struct NordDrop {
+    dev: Mutex<NordDropFFI>,
+    _cb: Box<CbContext>,
+    #[cfg(unix)]
+    fd_resolv: Mutex<Option<Box<FdResolverContext>>>,
+}
+
+#[cfg(unix)]
+struct FdResolverContext {
+    resolver: Box<dyn FdResolver>,
+}
+
+struct CbContext {
+    event_callback: Box<dyn EventCallback>,
+    key_store: Box<dyn KeyStore>,
+    logger: Box<dyn Logger>,
+}
+
+pub enum TransferDescriptor {
+    Path {
+        path: String,
+    },
+    Fd {
+        filename: String,
+        content_uri: String,
+        fd: Option<i32>,
+    },
+}
+
+pub enum TransferStateKind {
+    Cancel { by_peer: bool },
+    Failed { status: i64 },
+}
+
+pub struct TransferState {
+    pub created_at: SystemTime,
+    pub kind: TransferStateKind,
+}
+
+pub enum IncomingPathStateKind {
+    Pending { base_dir: String },
+    Started { bytes_received: i64 },
+    Failed { status: i64, bytes_received: i64 },
+    Completed { final_path: String },
+    Rejected { by_peer: bool, bytes_received: i64 },
+    Paused { bytes_received: i64 },
+}
+
+pub struct IncomingPathState {
+    pub created_at: SystemTime,
+    pub kind: IncomingPathStateKind,
+}
+
+pub struct IncomingPath {
+    pub file_id: String,
+    pub relative_path: String,
+    pub bytes: i64,
+    pub bytes_received: i64,
+    pub states: Vec<IncomingPathState>,
+}
+
+pub enum OutgoingPathStateKind {
+    Started { bytes_sent: i64 },
+    Failed { status: i64, bytes_sent: i64 },
+    Completed,
+    Rejected { by_peer: bool, bytes_sent: i64 },
+    Paused { bytes_sent: i64 },
+}
+
+pub struct OutgoingPathState {
+    pub created_at: SystemTime,
+    pub kind: OutgoingPathStateKind,
+}
+
+pub enum OutgoingFileSource {
+    BasePath { base_path: String },
+    ContentUri { uri: String },
+}
+
+pub struct OutgoingPath {
+    pub file_id: String,
+    pub relative_path: String,
+    pub bytes: i64,
+    pub bytes_sent: i64,
+    pub source: OutgoingFileSource,
+    pub states: Vec<OutgoingPathState>,
+}
+
+pub enum TransferKind {
+    Incoming { paths: Vec<IncomingPath> },
+    Outgoing { paths: Vec<OutgoingPath> },
+}
+
+pub struct TransferInfo {
+    pub id: String,
+    pub created_at: SystemTime,
+    pub peer: String,
+    pub states: Vec<TransferState>,
+    pub kind: TransferKind,
+}
+
+impl NordDrop {
+    pub fn new(
+        event_callback: Box<dyn EventCallback>,
+        key_store: Box<dyn KeyStore>,
+        logger: Box<dyn Logger>,
+    ) -> Result<Self> {
+        // TODO(msz): remove indirection
+
+        let mut cb = Box::new(CbContext {
+            event_callback,
+            key_store,
+            logger,
+        });
+
+        unsafe extern "C" fn call_on_event(ctx: *mut c_void, ev: *const c_char) {
```

Comment:
It looks weird to have `extern "C"` in here. I think we could get rid of all of them?

Reply:
I may have limited knowledge on this one, but `extern "C"` means "don't mangle this name as it must be ABI compatible". This is an inner function defined in `new` but it will still be not-mangled so this gives me a weird feeling - if it's exposed why have it defined inside of `new` and if it's intentional?

Also this raises a question.. Do we still need to do the whole extern "C" when we're using UniFFI?

Reply:
The `extern "C"` specifies the C ABI calling convention and isn't related to name mangling. There's `#[no_mangle]` attribute for that.
Of course, we don't need to use it. It is the same thing as with JSON - to be removed. That's why I was not bothered with making this code nice, since it will be removed in the subsequent PRs

Reply:
You're right. I mistook `extern` behaviour for C and C++. Thanks.